### ISSUE TITLE
copyiable iteration should suggest removing of `.copied()`

### ISSUE THREAD

**Issue by nyurik** [rust-lang/rust-clippy#13767]
### What it does

Currently, `cloned_instead_of_copied` lint suggests to replace `cloned()` with `copied()` in this example:

```rust
pub fn lengths(items: &[&str]) -> usize {
    items.iter().cloned().map(|one| one.len() + 1).sum()
}
```

but Rust allows `.cloned()` to be fully removed.  A lint should suggest that instead of suggesting `copied`.  I am not certain if this should be an extension to `cloned_instead_of_copied` or if its a new lint.

### Advantage

- Shorter, possibly faster code

### Example

```rust
#[must_use]
pub fn lengths(items: &[&str]) -> usize {
    items.iter().cloned().map(|one| one.len() + 1).sum()
}

fn main() {
    println!("Sum={}", lengths(&["hello", "world"]));
}
```

Could be written as:

```rust
pub fn lengths(items: &[&str]) -> usize {
    items.iter().map(|one| one.len() + 1).sum()
}
```

**Comment by DylanBulfin** [comment]
I was looking into this but think it's too complicated for me. Leaving a few notes here in case it helps anyone else.

I think there's a lot of context-specific logic that will have to be checked to see if it's removable. There are some existing lints that do similar things, mainly `unnecessary_to_owned` and `unnecessary_iter_cloned`

<details>

<summary>Here are a few example I used for testing</summary>

Here are a few examples that are superficially similar where `copied()` can't be removed:
```
fn testing() -> Box<dyn Iterator<Item = u64>> {
    Box::new([1u64, 2, 3, 4, 5].iter().copied())
}

fn testing2() -> Vec<u64> {
    [1u64, 2, 3, 4, 5].iter().copied().collect()
}

fn testing3(key: u32, vals: &HashMap<u32, u64>) -> u64 {
    if let Some(n) = vals.get(&key).copied() {
        n
    } else {
        0
    }
}
```

And here's another case where it *can* be removed, because of the inherent `impl<'a> Sum<&'a u64> for u64` implementation:
```
fn testing4() -> u64 {
    [1u64, 2, 3, 4, 5].iter().copied().sum()
}
```
</details>

**Comment by Jarcho** [comment]
Both `cloned` and `copied` change the item type from `&T` to `T` so this can't be done in general.
